// @finicity/connect-web-sdk/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CLOSE_POPUP_EVENT = exports.USER_EVENT = exports.ROUTE_EVENT = exports.WINDOW_EVENT = exports.PING_EVENT = exports.ERROR_EVENT = exports.DONE_EVENT = exports.URL_EVENT = exports.CANCEL_EVENT = exports.ACK_EVENT = exports.CONNECT_POPUP_HEIGHT = exports.CONNECT_POPUP_WIDTH = exports.POPUP_HEIGHT = exports.POPUP_WIDTH = exports.CONNECT_SDK_VERSION = exports.PLATFORM_POPUP = exports.PLATFORM_IFRAME = exports.STYLES_ID = exports.IFRAME_ID = void 0;
// NOTE: Changing this value should be considered a version breaking change!!!
var IFRAME_ID = 'finicityConnectIframe';
exports.IFRAME_ID = IFRAME_ID;
var STYLES_ID = 'finicityConnectStyles';
exports.STYLES_ID = STYLES_ID;
var PLATFORM_IFRAME = 'iframe';
exports.PLATFORM_IFRAME = PLATFORM_IFRAME;
var PLATFORM_POPUP = 'web';
exports.PLATFORM_POPUP = PLATFORM_POPUP;
var CONNECT_SDK_VERSION = '1.0.0-rc.4'; // NOTE: oauth popup window dimensions

exports.CONNECT_SDK_VERSION = CONNECT_SDK_VERSION;
var POPUP_WIDTH = 600;
exports.POPUP_WIDTH = POPUP_WIDTH;
var POPUP_HEIGHT = 600; // NOTE: connect popup window dimensions

exports.POPUP_HEIGHT = POPUP_HEIGHT;
var CONNECT_POPUP_WIDTH = 520;
exports.CONNECT_POPUP_WIDTH = CONNECT_POPUP_WIDTH;
var CONNECT_POPUP_HEIGHT = 720; // NOTE: available events

exports.CONNECT_POPUP_HEIGHT = CONNECT_POPUP_HEIGHT;
var ACK_EVENT = 'ack';
exports.ACK_EVENT = ACK_EVENT;
var CANCEL_EVENT = 'cancel';
exports.CANCEL_EVENT = CANCEL_EVENT;
var URL_EVENT = 'url';
exports.URL_EVENT = URL_EVENT;
var DONE_EVENT = 'done';
exports.DONE_EVENT = DONE_EVENT;
var ERROR_EVENT = 'error';
exports.ERROR_EVENT = ERROR_EVENT;
var PING_EVENT = 'ping';
exports.PING_EVENT = PING_EVENT;
var WINDOW_EVENT = 'window';
exports.WINDOW_EVENT = WINDOW_EVENT;
var ROUTE_EVENT = 'route';
exports.ROUTE_EVENT = ROUTE_EVENT;
var USER_EVENT = 'user';
exports.USER_EVENT = USER_EVENT;
var CLOSE_POPUP_EVENT = 'closePopup';
exports.CLOSE_POPUP_EVENT = CLOSE_POPUP_EVENT;
